internal/controller/clusterorder_controller.go

```diff
@@ -20,46 +20,247 @@ package controller
 
-	// TODO(user): your logic here
+	instance := &cloudkitv1alpha1.ClusterOrder{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 
-	return ctrl.Result{}, nil
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.handleUpdate(ctx, req, instance)
+	} else {
+		return r.handleDelete(ctx, req, instance)
+	}
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ClusterOrderReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	labelSelector := cloudkitClusterOrderNameLabel
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&cloudkitv1alpha1.ClusterOrder{}).
+		Watches(
+			&corev1.Namespace{},
+			handler.EnqueueRequestsFromMapFunc(r.mapObjectToCluster),
+			builder.WithPredicates(labelPredicate(labelSelector)),
+		).
+		Watches(
+			&corev1.ServiceAccount{},
+			handler.EnqueueRequestsFromMapFunc(r.mapObjectToCluster),
+			builder.WithPredicates(labelPredicate(labelSelector)),
+		).
+		Watches(
+			&rbacv1.RoleBinding{},
+			handler.EnqueueRequestsFromMapFunc(r.mapObjectToCluster),
+			builder.WithPredicates(labelPredicate(labelSelector)),
+		).
 		Complete(r)
 }
+
+func labelPredicate(label string) predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, exists := e.Object.GetLabels()[label]
+			return exists
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			_, oldExists := e.ObjectOld.GetLabels()[label]
+			_, newExists := e.ObjectNew.GetLabels()[label]
+			return oldExists || newExists
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			_, exists := e.Object.GetLabels()[label]
+			return exists
+		},
+	}
+}
+
+func (r *ClusterOrderReconciler) mapObjectToCluster(ctx context.Context, obj client.Object) []reconcile.Request {
+	log := ctrllog.FromContext(ctx)
+
+	clusterOrderName, exists := obj.GetLabels()[cloudkitClusterOrderNameLabel]
+	if !exists {
+		return nil
+	}
+
+	clusterOrderNamespace, exists := obj.GetLabels()[cloudkitClusterOrderNamespaceLabel]
+	if !exists {
+		return nil
+	}
+
+	log.Info("Selecting " + obj.GetName())
+
+	return []reconcile.Request{
+		{
+			NamespacedName: client.ObjectKey{
+				Name:      clusterOrderName,
+				Namespace: clusterOrderNamespace,
+			},
+		},
+	}
+}
+
+func (r *ClusterOrderReconciler) handleUpdate(ctx context.Context, _ ctrl.Request, instance *cloudkitv1alpha1.ClusterOrder) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Create or update " + instance.GetName())
+
+	// Do we have a finalizer yet?
+	if !controllerutil.ContainsFinalizer(instance, cloudkitFinalizer) {
+		controllerutil.AddFinalizer(instance, cloudkitFinalizer)
+		if err := r.Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Do we have a clusterReference yet?
+	if instance.Status.ClusterReference == nil {
+		log.Info("Adding cluster reference")
+		instance.Status.ClusterReference = newClusterReference()
```

Comment:
We should not add this until we have information to put inside it. That's the convention you'll see around the k8s API, and it's why we made sure this field is a pointer.

Reply:
That has actually already changed (https://github.com/larsks/cloudkit-operator/commit/d78276679aff6a7a4490e65343e2f75c0485d3ad#diff-c6507c1b735e1ac4b51a71ab5459385e88a41f5584de95efac80e0a90380429eL181).